pkg/webhook/controlplane/etcd.go

```diff
@@ -0,0 +1,143 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"fmt"
+	"github.com/gardener/gardener/pkg/operation/common"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GetBackupRestoreContainer returns an etcd backup-restore container with the given name, schedule, provider, image,
+// and additional provider-specific command line args and env variables.
+func GetBackupRestoreContainer(
+	name, schedule, storageProvider, image string,
+	args map[string]string,
+	env []corev1.EnvVar,
+	volumeMounts []corev1.VolumeMount,
+) *corev1.Container {
+	// Determine the storage provider
+	// A non-empty storage provider is used only for the etcd-main stateful set
+	provider := ""
+	if name == common.EtcdMainStatefulSetName {
+		provider = storageProvider
+	}
+
+	c := &corev1.Container{
+		Name: "backup-restore",
+		Command: []string{
+			"etcdbrctl",
+			"server",
+			fmt.Sprintf("--schedule=%s", schedule),
+			"--data-dir=/var/etcd/data/new.etcd",
+			fmt.Sprintf("--storage-provider=%s", provider),
+			fmt.Sprintf("--store-prefix=%s", name),
+			"--cert=/var/etcd/ssl/client/tls.crt",
+			"--key=/var/etcd/ssl/client/tls.key",
+			"--cacert=/var/etcd/ssl/ca/ca.crt",
+			"--insecure-transport=false",
+			"--insecure-skip-tls-verify=false",
+			fmt.Sprintf("--endpoints=https://%s-0:2379", name),
+			"--etcd-connection-timeout=300",
+			"--delta-snapshot-period-seconds=300",
+			"--delta-snapshot-memory-limit=104857600", // 100MB
+			"--garbage-collection-period-seconds=43200",
+			"--snapstore-temp-directory=/var/etcd/data/temp",
+		},
+		Env:             []corev1.EnvVar{},
+		Image:           image,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          "server",
+				ContainerPort: 8080,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		},
+		Resources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("100m"),
+				corev1.ResourceMemory: resource.MustParse("128Mi"),
+			},
+			Limits: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("500m"),
+				corev1.ResourceMemory: resource.MustParse("2Gi"),
+			},
+		},
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      name,
+				MountPath: "/var/etcd/data",
+			},
+			{
+				Name:      "ca-etcd",
+				MountPath: "/var/etcd/ssl/ca",
+			},
+			{
+				Name:      "etcd-client-tls",
+				MountPath: "/var/etcd/ssl/client",
+			},
+		},
+	}
+
+	// Ensure additional command line args
+	for k, v := range args {
+		c.Command = EnsureStringWithPrefix(c.Command, fmt.Sprintf("--%s=", k), v)
+	}
+
+	// Ensure additional env variables
+	for _, envVar := range env {
+		c.Env = EnsureEnvVarWithName(c.Env, envVar)
```

Comment:
Should we not add any env variables for backup if we are mutating the etcd-events statefulset? The empty provider flag should already prevent the sidecar from taking backups, so it would be little cleaner to not add the configuration (= env variables).

Reply:
Is it certain that all env variables are only needed if backup is enabled? If yes, this is a good idea. If not, then the logic for checking the "role" should be moved to the invoking code, which would be a bigger change.

Reply:
Yes, the env variables are only needed if backup is enabled. That's how it works today.

Reply:
Still, I moved the role checking to the ensurers, I believe it's cleaner this way. I also added tests for both etcd-main and etcd-events.